### PR TITLE
[1.0.x] Support bulkhead group

### PR DESCRIPTION
**Pull request by Ferioney**
Related to [Spring Cloud OpenFeign Issue 514.](https://github.com/spring-cloud/spring-cloud-openfeign/issues/514)
These changes required the latest version of Spring Cloud Commons.

### CODE REVIEW

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java [hunk 1]
```diff
@@ -78,6 +81,26 @@ public Resilience4JCircuitBreaker(String id,
...
+		this.circuitBreakerConfig = circuitBreakerConfig;
+		this.registry = circuitBreakerRegistry;
+		this.timeLimiterRegistry = timeLimiterRegistry;
+		this.timeLimiterConfig = timeLimiterConfig;
+		this.executorService = executorService;
+		this.circuitBreakerCustomizer = circuitBreakerCustomizer;
+		this.bulkheadProvider = bulkheadProvider;
+	}
+
+	public Resilience4JCircuitBreaker(String id,
```
Comment (ryanjbaxter) [inline spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java:94]: Can we have the old constructor call this one?

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java [hunk 2]
```diff
@@ -114,6 +121,18 @@ public Resilience4JCircuitBreaker create(String id) {
 				Optional.ofNullable(circuitBreakerCustomizers.get(id)), bulkheadProvider);
 	}
 
+	@Override
+	public Resilience4JCircuitBreaker create(String id, String groupName) {
```
Comment (ryanjbaxter) [inline spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java:125]: Could we find a way to reuse code between the two create methods?

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java (hunk 1 above)
Comment (Ferioney) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java:88]: Yep, changed

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java (hunk 2 above)
Comment (Ferioney) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java:120]: Changed

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java [hunk 3]
```diff
@@ -70,14 +73,28 @@ public Resilience4JCircuitBreaker(String id,
...
+									  String groupName,
+									  io.github.resilience4j.circuitbreaker.CircuitBreakerConfig circuitBreakerConfig,
+									  TimeLimiterConfig timeLimiterConfig,
+									  CircuitBreakerRegistry circuitBreakerRegistry,
+									  TimeLimiterRegistry timeLimiterRegistry, ExecutorService executorService,
+									  Optional<Customizer<io.github.resilience4j.circuitbreaker.CircuitBreaker>> circuitBreakerCustomizer,
+									  Resilience4jBulkheadProvider bulkheadProvider) {
 		this.id = id;
+		this.groupName = groupName;
```
Comment (ryanjbaxter) [inline spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java:97]: Can this be leveraged in the non-bulkhead case?  Maybe as a tag?

**Review by ryanjbaxter — CHANGES_REQUESTED** [review]

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java [hunk 4]
```diff
@@ -103,15 +106,22 @@ public void configureExecutorService(ExecutorService executorService) {
 		this.executorService = executorService;
 	}
 
+	public void configureExecutorServices(ConcurrentHashMap<String, ExecutorService> executorServices) {
+		this.executorServices = executorServices;
+	}
+
 	@Override
 	public Resilience4JCircuitBreaker create(String id) {
 		Assert.hasText(id, "A CircuitBreaker must have an id.");
```
[the remark this answers is not in the document]
Comment (Ferioney) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java:111]: Removed

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java (hunk 3 above)
Comment (Ferioney) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreaker.java:99]: Added tags with group

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ryanjbaxter) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java:111]: Its still present

**Review by ryanjbaxter — CHANGES_REQUESTED** [review]

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ferioney) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java:111]: I don't see `configureExecutorServices(ConcurrentHashMap<String, ExecutorService>)` method after the last changes

[the remark this answers is not in the document]
Comment (ryanjbaxter) [reply spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4JCircuitBreakerFactory.java:111]: My mistake I thought this was calling the other public `create` method, not the private one.

**Pull request merged**